reconcile/saas_auto_promotions_manager/integration.py

```diff
@@ -0,0 +1,223 @@
+
+
+def get_commit_sha(repo_url: str, ref: str, github: Github, gitlab: GitLabApi) -> str:
+    if is_commit_sha(ref):
+        # This is already a commit sha - no need to query VCS
+        return ref
+    if "github.com" in repo_url:
+        repo_name = repo_url.rstrip("/").replace("https://github.com/", "")
+        gh = github.get_repo(repo_name)
+        commit = gh.get_commit(sha=ref)
+        return commit.sha
+    if "gitlab.cee" in repo_url:
+        project = gitlab.get_project(repo_url)
+        commits = project.commits.list(ref_name=ref)
+        return commits[0].id
+    raise RuntimeError(f"Unsupported Repo URL {repo_url}")
+
+
+@dataclass
+class PublisherCurrentState:
+    commit_sha: str
+    config_hash: Optional[str]
+    channel_name: str
+    subscribers: list[SubscriberTarget]
+
+
+def states_with_auto_promotions_and_successful_deployment(
+    saas_channels: Iterable[SaaSChannel],
+    gitlab: GitLabApi,
+    github_default_token: str,
+    secret_reader: SecretReaderBase,
+    deployment_state: State,
+) -> list[PublisherCurrentState]:
+    ans = []
+    deployed_commits_by_channel = group_deployed_commits_by_channel(
+        commits=deployment_state.ls()
+    )
+    commit_sha_cache: dict[str, str] = {}
+    deployment_info_cache: dict[str, dict[str, Any]] = {}
+    for channel in saas_channels:
+        has_auto_promotion = False
+        for sub in channel.subscribers:
+            has_auto_promotion |= sub.auto_promote or False
+        if not has_auto_promotion:
+            continue
+        cache_key = (
+            f"{channel.name}/{channel.publisher.repo_url}/{channel.publisher.ref}"
+        )
+        if cache_key not in commit_sha_cache:
+            github = init_github(
+                auth_code=channel.publisher.auth_code,
+                secret_reader=secret_reader,
+                default_token=github_default_token,
+            )
+            commit_sha_cache[cache_key] = get_commit_sha(
+                repo_url=channel.publisher.repo_url,
+                ref=channel.publisher.ref,
+                github=github,
+                gitlab=gitlab,
+            )
+        publisher_commit_sha = commit_sha_cache[cache_key]
+        if publisher_commit_sha not in deployed_commits_by_channel[channel.name]:
+            print(
+                f"[{channel.name}] Commit sha {publisher_commit_sha} not found in deployments"
+            )
+            continue
+        if cache_key not in deployment_info_cache:
+            deployment_info_cache[cache_key] = deployment_state.get(
+                f"promotions/{channel.name}/{publisher_commit_sha}"
+            )
+        deployment_info = deployment_info_cache[cache_key]
+        if not deployment_info.get("success"):
+            print(
+                f"[{channel.name}] Commit sha {publisher_commit_sha} wasnt successfully deployed"
+            )
+            continue
+
+        ans.append(
+            PublisherCurrentState(
```

Comment:
I think I would revert this logic to returning a list of `SubscriberCurrentState` to be able to properly handle multiple subscription (APPSRE-7307)

Reply:
Makes a lot of sense :+1: I will make it Subscriber centric once I start working on it properly again (once the design doc is merged).